lib/fd.js

```diff
@@ -0,0 +1,309 @@
+const fs = require('fs')
+const pump = require('pump')
+const errors = require('./errors')
+const byteStream = require('byte-stream')
+const through = require('through2')
+const pumpify = require('pumpify')
+const messages = require('./messages')
+
+const noop = () => {}
+
+const {
+  O_RDONLY,
+  O_WRONLY,
+  O_RDWR,
+  O_CREAT,
+  O_TRUNC,
+  O_APPEND,
+  O_SYNC,
+  O_EXCL
+} = fs.constants
+const O_ACCMODE = 3
+
+class FileDescriptor {
+  constructor (drive, path, stat, readable, writable, appending, creating) {
+    this.drive = drive
+    this.stat = stat
+    this.path = path
+
+    this.readable = readable
+    this.writable = writable
+    this.creating = creating
+    this.appending = appending
+
+    this.position = null
+    this.blockPosition = stat ? stat.offset : null
+    this.blockOffset = 0
+
+    this._err = null
+    if (this.writable) {
+      if (this.appending) {
+        this._appendStream = this.drive.createReadStream(this.path)
+        this.position = this.stat.size
+      }
+    }
+
+    this._batcher = byteStream({ time: 100, limit: 4096 * 16 })
+
+    this._range = null
+  }
+
+  read (buffer, offset, len, pos, cb) {
+    if (!this.readable) return cb(new errors.BadFileDescriptor('File descriptor not open for reading.'))
+    if (this.position !== null && this.position === pos) this._read(buffer, offset, len, cb)
+    else this._seekAndRead(buffer, offset, len, pos, cb)
+  }
+
+  write (buffer, offset, len, pos, cb) {
+    if (!this.writable) return cb(new errors.BadFileDescriptor('File descriptor not open for writing.'))
+    if (!this.stat && !this.creating) {
+      return cb(new errors.BadFileDescriptor('File descriptor not open in create mode.'))
+    }
+    if (this.position !== null && pos !== this.position) {
+      return cb(new errors.BadFileDescriptor('Random-access writes are not currently supported.'))
+    }
+    if (this.appending && pos < this.stat.size) {
+      return cb(new errors.BadFileDescriptor('Position cannot be less than the file size when appending.'))
+    }
+    if (!this._writeStream && !this.appending && pos) {
+      return cb(new errors.BadFileDescriptor('Random-access writes are not currently supported.'))
```

Comment:
for all these sync calls to the cb above, next tick them, `process.nextTick(cb, new errors...)`